File: Components/WinForms/VB/MVVMMainDemo/Modules/Navigation/TabbedFormModule.vb
Imports System
Imports DevExpress.MVVM.Demos.ViewModels
Imports DevExpress.Utils.MVVM.Services

Namespace DevExpress.MVVM.Demos.Navigation
    Partial Public Class TabbedFormModule
        Inherits DevExpress.XtraBars.TabForm

        Public Sub New()
            InitializeComponent()
            If Not mvvmContext.IsDesignMode Then
                InitializeNavigation()
            End If
        End Sub
        Private Sub InitializeNavigation()
            ' We want the DocumentManagerService based on TabFormControl to be a navigation provider
            mvvmContext.RegisterService(DocumentManagerService.Create(TabFormControl))
            ' We want to use buttons in TabFormControl to show the specific modules
            Dim fluentAPI = mvvmContext.OfType(Of ExpensesDbContextViewModel)()
            fluentAPI.BindCommand(biAccounts, Sub(x, m) x.Show(m), Function(x) x.Modules(0))
            fluentAPI.BindCommand(biCategories, Sub(x, m) x.Show(m), Function(x) x.Modules(1))
            fluentAPI.BindCommand(biTransactions, Sub(x, m) x.Show(m), Function(x) x.Modules(2))
            ' We want show the default module when our TabForm is loaded
            fluentAPI.WithEvent(Of EventArgs)(Me, "Load").
                EventToCommand(Sub(x) x.OnLoaded(Nothing), Function(x) x.DefaultModule)
        End Sub
    End Class
#Region "Module Starter"
    Partial Public Class TabbedFormModuleStarter
        Inherits TutorialControl

        Private sourceCode As DXperience.Demos.CodeDemo.RichEditUserControl
        Private showModuleButton As DevExpress.XtraEditors.PanelControl
        Public Sub New()
            InitializeComponent()
            Dim asm = GetType(DevExpress.Tutorials.ucAboutPage).Assembly
            AddHandler CType(New DevExpress.Tutorials.Controls.OverviewButton(showModuleButton,
                DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonNormal.png", asm),
                DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonHover.png", asm),
                DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonPressed.png", asm), String.Empty), 
            DevExpress.Tutorials.Controls.OverviewButton).ButtonClick, AddressOf showModuleButton_Click
            '
            LoadSourceCode(sourceCode, "InitializeNavigation", GetType(TabbedFormModule), True)
        End Sub
        Private Sub InitializeComponent()
            Me.sourceCode = New DevExpress.DXperience.Demos.CodeDemo.RichEditUserControl()
            Me.showModuleButton = New XtraEditors.PanelControl()
            Me.SuspendLayout()
            ' 
            ' showModuleButton
            ' 
            Me.showModuleButton.BorderStyle = XtraEditors.Controls.BorderStyles.NoBorder
            Me.showModuleButton.Dock = System.Windows.Forms.DockStyle.Fill
            Me.showModuleButton.Name = "showModuleButton"
            Me.showModuleButton.Text = "Start the TabbedForm Module"
            ' 
            ' sourceCode
            ' 
            Me.sourceCode.Dock = System.Windows.Forms.DockStyle.Bottom
            Me.sourceCode.Location = New System.Drawing.Point(5, 217)
            Me.sourceCode.Margin = New System.Windows.Forms.Padding(0)
            Me.sourceCode.Name = "sourceCode"
            Me.sourceCode.Size = New System.Drawing.Size(779, 210)
            ' 
            ' WindowedModule
            ' 
            Me.AutoScaleDimensions = New System.Drawing.SizeF(6.0F, 13.0F)
            Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font
            Me.BorderStyle = System.Windows.Forms.BorderStyle.None
            Me.Controls.Add(Me.showModuleButton)
            Me.Controls.Add(Me.sourceCode)
            Me.Name = "TabbedFormModuleStarter"
            Me.ResumeLayout(False)
            Me.PerformLayout()
        End Sub
        '
        Private [module] As TabbedFormModule
        Private Sub showModuleButton_Click(ByVal sender As Object, ByVal e As EventArgs)
            If [module] Is Nothing Then
                [module] = New TabbedFormModule()
                [module].StartPosition = System.Windows.Forms.FormStartPosition.Manual
                If ParentFormMain IsNot Nothing Then
                    [module].Location = New System.Drawing.Point(ParentFormMain.Left + (ParentFormMain.Width - [module].Width) \ 2, ParentFormMain.Top + (ParentFormMain.Height - [module].Height) \ 2)
                End If
                AddHandler [module].FormClosed, AddressOf activeModule_FormClosed
            End If
            showModuleButton.Enabled = False
            [module].Show(Me)
        End Sub
        Private Sub activeModule_FormClosed(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosedEventArgs)
            If [module] IsNot Nothing Then
                RemoveHandler [module].FormClosed, AddressOf activeModule_FormClosed
            End If
            [module] = Nothing
            showModuleButton.Enabled = True
        End Sub
        Protected Overrides Sub DoVisibleChanged(ByVal visible As Boolean)
            MyBase.DoVisibleChanged(visible)
            If (Not visible) AndAlso [module] IsNot Nothing AndAlso [module].Visible Then
                [module].Close()
            End If
        End Sub
    End Class
#End Region
End Namespace

File: Components/WinForms/CS/MVVMMainDemo/CodeExamples.Resources/TabbedFormModule.cs
using System;
using DevExpress.MVVM.Demos.ViewModels;
using DevExpress.Utils.MVVM.Services;

namespace DevExpress.MVVM.Demos.Navigation {
    public partial class TabbedFormModule : DevExpress.XtraBars.TabForm {
        public TabbedFormModule() {
            InitializeComponent();
            if(!mvvmContext.IsDesignMode)
                InitializeNavigation(true);
        }
        public TabbedFormModule(bool loadDefaultModule) {
            InitializeComponent();
            if(!mvvmContext.IsDesignMode)
                InitializeNavigation(loadDefaultModule);
        }
        void tabFormControl_OuterFormCreating(object sender, XtraBars.OuterFormCreatingEventArgs e) {
            // creating secondary Tab Form
            e.Form = new TabbedFormModule(false);
        }
        //
        void InitializeNavigation(bool loadDefaultModule) {
            // We want the DocumentManagerService based on TabFormControl to be a navigation provider
            mvvmContext.RegisterService(DocumentManagerService.Create(tabFormControl));
            // We want to use buttons in TabFormControl to show the specific modules
            var fluentAPI = mvvmContext.OfType<ExpensesDbContextViewModel>();
            // assign True if needed to load default module (True for main Tab Form, False for secondary Tab Form)
            fluentAPI.ViewModel.AllowDefaultModule = loadDefaultModule;
            fluentAPI.BindCommand(biAccounts, (x, m) => x.Show(m), x => x.Modules[0]);
            fluentAPI.BindCommand(biCategories, (x, m) => x.Show(m), x => x.Modules[1]);
            fluentAPI.BindCommand(biTransactions, (x, m) => x.Show(m), x => x.Modules[2]);
            // We want show the default module when our TabForm is loaded
            fluentAPI.WithEvent<EventArgs>(this, "Load")
                .EventToCommand(x => x.OnLoaded(null), x => x.DefaultModule);
            // We want to add a new Entity into the active Entities list
            fluentAPI.BindCommand(tabFormControl.AddPageButton, x => x.AddNew());
        }
    }
    #region Module Starter
    public partial class TabbedFormModuleStarter : TutorialControl {
        DXperience.Demos.CodeDemo.RichEditUserControl sourceCode;
        DevExpress.XtraEditors.PanelControl showModuleButton;
        public TabbedFormModuleStarter() {
            InitializeComponent();
            var asm = typeof(DevExpress.Tutorials.ucAboutPage).Assembly;
            new DevExpress.Tutorials.Controls.OverviewButton(showModuleButton,
                    DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonNormal.png", asm),
                    DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonHover.png", asm),
                    DevExpress.Utils.ResourceImageHelper.CreateImageFromResources("DevExpress.Tutorials.Images.RunButtonPressed.png", asm),
                    string.Empty).ButtonClick += showModuleButton_Click;
            //
            LoadSourceCode(sourceCode, "tabFormControl_OuterFormCreating", typeof(TabbedFormModule));
            LoadSourceCode(sourceCode, "InitializeNavigation", typeof(TabbedFormModule), true);
        }
        void InitializeComponent() {
            this.sourceCode = new DevExpress.DXperience.Demos.CodeDemo.RichEditUserControl();
            this.showModuleButton = new XtraEditors.PanelControl();
            this.SuspendLayout();
            // 
            // showModuleButton
            // 
            this.showModuleButton.BorderStyle = XtraEditors.Controls.BorderStyles.NoBorder;
            this.showModuleButton.Dock = System.Windows.Forms.DockStyle.Fill;
            this.showModuleButton.Name = "showModuleButton";
            this.showModuleButton.Text = "Start the TabbedForm Module";
            // 
            // sourceCode
            // 
            this.sourceCode.Dock = System.Windows.Forms.DockStyle.Bottom;
            this.sourceCode.Location = new System.Drawing.Point(5, 217);
            this.sourceCode.Margin = new System.Windows.Forms.Padding(0);
            this.sourceCode.Name = "sourceCode";
            this.sourceCode.Size = new System.Drawing.Size(779, 352);
            // 
            // WindowedModule
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.BorderStyle = System.Windows.Forms.BorderStyle.None;
            this.Controls.Add(this.showModuleButton);
            this.Controls.Add(this.sourceCode);
            this.Name = "TabbedFormModuleStarter";
            this.ResumeLayout(false);
            this.PerformLayout();
        }
        //
        TabbedFormModule module;
        void showModuleButton_Click(object sender, EventArgs e) {
            if(module == null) {
                module = new TabbedFormModule();
                module.StartPosition = System.Windows.Forms.FormStartPosition.Manual;
                if(ParentFormMain != null) {
                    module.Location = new System.Drawing.Point(
                        ParentFormMain.Left + (ParentFormMain.Width - module.Width) / 2,
                        ParentFormMain.Top + (ParentFormMain.Height - module.Height) / 2);
                }
                module.FormClosed += activeModule_FormClosed;
            }
            showModuleButton.Enabled = false;
            module.Show(this);
        }
        void activeModule_FormClosed(object sender, System.Windows.Forms.FormClosedEventArgs e) {
            if(module != null)
                module.FormClosed -= activeModule_FormClosed;
            module = null;
            showModuleButton.Enabled = true;
        }
        protected override void DoVisibleChanged(bool visible) {
            base.DoVisibleChanged(visible);
            if(!visible && module != null && module.Visible)
                module.Close();
        }
    }
    #endregion
}
